### PR TITLE
Add reverseproxy role (and a few small fixes)

### PR DESCRIPTION
**Pull request by ubergeek42**
**Note**: This PR has been rewritten to use nginx as a reverse proxy to kattis instead of squid.

This adds a new role for a reverseproxy server (assuming it's `{{ baseip }}.205`, there's a duplicate entry for it in the `script/build_host` file as `.214`.

I fixed a couple small issues, and changed the linter to be silent instead of warn for the couple things we don't care about.

I put the following content in `group_vars/all/dynamic.yml`, and testing on a "team" machine this seems to do the right thing. I can access kattis, and it seems to work as expected, but nothing else. We'll still need firewall rules to make sure the teams can only access the reverse proxy server, but that should be config-wise identical to how they would access a local kattis instance (the only difference between the reverse proxy server has internet access).

```yaml
reverseproxy_key: |
  CONTENT OF THE SSL CERTFILE
reverseproxy_cert: |
  CONTENT OF THE SSL KEYFILE
reverseproxy_sites:
  nac22.kattis.com:
    - {path: "/"}
  fonts.gstatic.com:
    - {path: "/"}
  fonts.googleapis.com:
    - {path: "/css", with_args: true, exact: true}
  ajax.googleapis.com:
    - {path: "/ajax/libs/jquery/3.5.1/jquery.min.js", exact: true}
    - {path: "/ajax/libs/jqueryui/1.12.1/jquery-ui.min.js", exact: true}
    - {path: "/ajax/libs/jqueryui/1.12.1/themes/base/jquery-ui.min.css", exact: true}
  cdnjs.cloudflare.com:
    - {path: "/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css", exact: true}
    - {path: "/ajax/libs/font-awesome/4.7.0/fonts/fontawesome-webfont.woff2", exact: true}
    - {path: "/ajax/libs/moment.js/2.24.0/moment.min.js", exact: true}
    - {path: "/ajax/libs/raphael/2.2.8/raphael.min.js", exact: true}
    - {path: "/ajax/libs/select2/3.5.4/select2.min.js", exact: true}
    - {path: "/ajax/libs/select2/3.5.4/select2.css", exact: true}
    - {path: "/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js", exact: true}
    - {path: "/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css", exact: true}
    - {path: "/ajax/libs/bootstrap-datetimepicker/4.17.47/js/bootstrap-datetimepicker.min.js", exact: true}
    - {path: "/ajax/libs/bootstrap-datetimepicker/4.17.47/css/bootstrap-datetimepicker.min.css", exact: true}
  cdn.jsdelivr.net:
    - {path: "/npm/daterangepicker/daterangepicker.min.js", exact: true}
    - {path: "/npm/daterangepicker/daterangepicker.css", exact: true}
#   # These two are probably analytics/metrics/monitoring stuff that seem unimportant
#   static.cloudflareinsights.com:
#     - {path: "/beacon.min.js"}
#   static.site24x7rum.eu:
#     - {path: "/beacon/site24x7rum-min.js", with_args: true, exact: true}
#   # The next entries are for the orange help widget on kattis.com
#   euc-widget.freshworks.com:
#     - {path: "/widgets/79000000133.js", exact: true}
#     - {path: "/widgets/79000000133.json", exact: true, with_args: true, }
#     - {path: "/widgetBase"}
#   kattis.freshdesk.com: [{path: "/"}]
#   support.kattis.com: [{path: "/"}]
```

### CODE REVIEW

**Comment by troy2914** [comment]
The 205 is the right entry for the squid server.

**Comment by troy2914** [comment]
and yes, we have inserted the ICPC keys into the firefox CA keystore with
```
/usr/lib/firefox/distribution/policies.json
```
containing
```
{
         "policies" :{
                 "Certificates" : {
                         "ImportEnterpriseRoots": true,
                         "Install": [
                                 "ICPC_WF_Root_CA.crt", "/usr/share/ca-certificates/ICPC_WF_Root_CA.crt", "ICPC_WF_Intermediate_CA.crt", "/usr/share/ca-certificates/ICPC_WF_Intermediate_CA.crt"
                         ]
                 }
         }
}
```

File: roles/squidserver/tasks/main.yml [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+      - squid
+
+- name: configure squid
+  ansible.builtin.template:
+    dest: /etc/squid/squid.conf
+    src: templates/squid.conf.j2
+    mode: 0644
+    owner: root
+    group: root
+
```
Comment (vmcj) [inline roles/squidserver/tasks/main.yml:15]:
```suggestion
```

**Comment by troy2914** [comment]
> Ansible code looks fine to me., just to be sure, if there would be a `kattis submitclient` it will also use this right?

lol,  nobody has told us to install a cmdline tool.

**Comment by vmcj** [comment]
> > Ansible code looks fine to me., just to be sure, if there would be a `kattis submitclient` it will also use this right?
> 
> lol, nobody has told us to install a cmdline tool.

So lets assume that will just work.

**Comment by troy2914** [comment]
> > > Ansible code looks fine to me., just to be sure, if there would be a `kattis submitclient` it will also use this right?
> > 
> > 
> > lol, nobody has told us to install a cmdline tool.
> 
> So lets assume that will just work.

in the past we have built perteam foo for kattis and it included the url, ex:
```
[kattis]
loginurl: https://kattis/login
submissionurl: https://kattis/judge_upload
```

so shouldn't be an issue.

**Comment by ubergeek42** [comment]
@troy2914 @vmcj I rewrote this PR to be an nginx server acting as a reverse proxy instead of a squid server, which gives us much finer grained control over the network resources we proxy. Give it a look?

File: lastminute.yml [hunk 2]
```diff
@@ -21,6 +22,13 @@
     - role: ldapserver
       tags: ldapserver
 
+- name: configure the squid server
+  hosts: squidserver
```
Comment (vmcj) [inline lastminute.yml:26]: Is this now the wrong groupname?

File: roles/hyperthreadingdisabled/tasks/main.yml [hunk 3]
```diff
@@ -1,13 +1,10 @@
 # Download latest version from DOMjudge repo
 # Alternative is to put the correct file in /tmp and set internet to false
-- name: Download turboboost script
+- name: Download turboboost script  # noqa: risky-file-permissions
```
Comment (vmcj) [inline roles/hyperthreadingdisabled/tasks/main.yml:3]:
```suggestion
- name: Download turboboost script
```

File: roles/hyperthreadingdisabled/tasks/main.yml [hunk 4]
```diff
@@ -1,13 +1,10 @@
...
   delegate_to: localhost
   ansible.builtin.get_url:
     url: https://raw.githubusercontent.com/DOMjudge/domjudge-scripts/main/provision-contest/disable-turboboost_ht
     dest: /tmp/turboboost_ht
-    mode: 0777
```
Comment (vmcj) [inline roles/hyperthreadingdisabled/tasks/main.yml:8]: This one can stay, and than the problem is fixed, the {user,group} is not checked in the risky-filepermissions

File: roles/reverseproxy/tasks/main.yml [hunk 5]
```diff
@@ -0,0 +1,8 @@
+---
+- name: server configuration
+  include_tasks: server.yml
+  when: reverseproxy_server is true
```
Comment (vmcj) [inline roles/reverseproxy/tasks/main.yml:4]:
```suggestion
  when: reverseproxy_server
```

File: roles/reverseproxy/tasks/main.yml [hunk 6]
```diff
@@ -0,0 +1,8 @@
...
+
+- name: client configuration
+  include_tasks: client.yml
+  when: reverseproxy_server is false
```
Comment (vmcj) [inline roles/reverseproxy/tasks/main.yml:8]:
```suggestion
  when: not reverseproxy_server
```

Might be personal preference though

File: roles/reverseproxy/tasks/server.yml [hunk 7]
```diff
@@ -0,0 +1,39 @@
+---
+- name: install nginx
+  ansible.builtin.apt: pkg=nginx state=present
```
Comment (vmcj) [inline roles/reverseproxy/tasks/server.yml:3]:
```suggestion
  ansible.builtin.apt:
    pkg: nginx
    state: present
```
Might be personal preference although this blog article (https://www.ansible.com/blog/ansible-best-practices-essentials) agrees with me. Both will work.

File: roles/reverseproxy/tasks/server.yml [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+  notify: reload nginx
+
+- name: install the ssl key
+  ansible.builtin.copy:
+    content: "{{ reverseproxy_key }}"
+    dest: /etc/ssl/reverseproxy.key
+    mode: 0600
+
+- name: install the ssl certificate
+  ansible.builtin.copy:
```
Comment (vmcj) [inline roles/reverseproxy/tasks/server.yml:18]: I think a loop works better here to bundle all cert related things together (for readability)

File: roles/reverseproxy/templates/nginx.conf.j2 [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+        ssl_certificate_key /etc/ssl/reverseproxy.key;
+
+        # Tune some buffer settings
+        proxy_buffer_size 16k;
+        proxy_busy_buffers_size 24k;
+        proxy_buffers 64 4k;
+
+        # Set the hostname as a variable, this forces nginx to do a dns lookup
+        # otherwise it only looks it up when it starts and never refreshes it
+        set $server_host "{{server}}";
```
Comment (vmcj) [inline roles/reverseproxy/templates/nginx.conf.j2:19]: Why is this needed?

File: roles/reverseproxy/templates/nginx.conf.j2 [hunk 10]
```diff
@@ -0,0 +1,38 @@
...
+
+        {% for item in urls %}
+        location {% if 'exact' in item and item.exact %}={% endif %} {{ item.path }} {
```
Comment (vmcj) [inline roles/reverseproxy/templates/nginx.conf.j2:22]: I think "convention" is:
```suggestion
        location {% if item.exact is defined and item.exact %}={% endif %} {{ item.path }} {
```
but I haven't tested it,

**Review by vmcj — APPROVED** [review]
Some comments but I think the book will work and can be merged + used.

File: roles/reverseproxy/tasks/main.yml (hunk 5 above)
Comment (troy2914) [reply roles/reverseproxy/tasks/main.yml:4]: i agree with this change

File: lastminute.yml (hunk 2 above)
Comment (troy2914) [reply lastminute.yml:26]: should we also rename that 205 host?  to what?

File: roles/reverseproxy/templates/nginx.conf.j2 (hunk 9 above)
Comment (ubergeek42) [reply roles/reverseproxy/templates/nginx.conf.j2:19]: Because the resources are mostly hosted on CDNs, and they typically have short TTLs on their DNS entries/might rotate an IP out of service during the contest. If nginx only looks it up on start, then if the dns entry changes/the original IP stops working, nginx will never look it up again.

File: roles/reverseproxy/tasks/server.yml (hunk 7 above)
Comment (ubergeek42) [reply roles/reverseproxy/tasks/server.yml:3]: I tend to do it out of habit for single packages/to keep the playbooks shorter if the arguments are simple, but I'm happy to follow best practices (and the arguments for it make plenty of sense)
